node/src/coordinator.rs

```diff
@@ -317,29 +342,155 @@ impl Coordinator {
         let (network_client, mut channel_receiver, _handle) =
             run_network_client(Arc::new(sender), Box::new(receiver));
 
-        let key = EcdsaSignatureProvider::run_key_generation_client(
-            mpc_config,
-            network_client,
-            &mut channel_receiver,
-        )
-        .await?;
-
-        keyshare_storage
-            .store_key(Keyshare::from_legacy(&LegacyRootKeyshareData {
-                epoch: 0,
-                private_share: key.private_share,
-                public_key: key.public_key,
-            }))
-            .await?;
-        let my_public_key = affine_point_to_public_key(key.public_key)?;
-        chain_txn_sender
-            .send(ChainSendTransactionRequest::VotePk(ChainVotePkArgs {
-                public_key: my_public_key,
-            }))
-            .await?;
-
-        // Just halt and wait for the running state.
-        Ok(MpcJobResult::HaltUntilInterrupted)
+        let is_leader = mpc_config.is_leader_for_keygen();
+        let keyshare_storage = Arc::new(keyshare_storage);
+        if !is_leader {
+            'follower: loop {
```

Comment:
I think this needs to be extracted to a separate function, so that when any of the `?`s errors out, we can loop again instead of exiting out of the whole thing.

(btw this should live in a separate file I think; it's not really the coordinator's job)